test/test_jit.py

```diff
@@ -89,10 +93,11 @@ def test_function_as_argument(self):
         cx = Variable(torch.randn(3, 20))
         lstm = nn.LSTMCell(10, 20)
 
-        def a_function(a, b):
+        @torch.jit.trace(nderivs=0, params=lstm.parameters())
```

Comment:
Following this line of reasoning to its conclusion, the suggestion is that any time we encounter a Variable which is not an input and not a result of a traced computation, instead of turning it into a constant (as is the behavior today, but @apaszke has said they're not too keen on this behavior so it could be changed), we instead *add* another input to the compiler IR we a tracing, and *furthermore* retain a reference to a Variable on this input. If we don't have a reference to the Variable, when someone wants to interpret the trace, they wouldn't be able to find a value to fill in for the variable.

I'm not sure if this is good design, but it has some good properties (user no longer has to specify parameters for functions.) There is some work we have to do to implement this correctly. One thing that is dangerous is that if a user *allocates* a Variable while within tracing, we MUST NOT treat this as an input. So moving into this regime will force us to teach the Variable allocation process to know something about whether or not we are tracing or not. Additionally, we probably don't want this logic to trigger when we are tracing backwards, so it has to be conditional on that as well. It's hard to say exactly how complicated implementing this would be, but I know at the very least these two things would have to be addressed.